### PR TITLE
Current main

### PR DESCRIPTION
**Pull request by andrew2003**

### CODE REVIEW

File: .DS_Store
Comment (zk-steve) [inline .DS_Store:1]: don't commit this file. apply for any similar files

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+/target
+.idea
+Cargo.lock
```
Comment (zk-steve) [inline .gitignore]: remove

File: .dockerignore [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# Ignore target directory generated by Cargo
+target/
+
+# Ignore Rust build artifacts
+Cargo.lock
```
Comment (zk-steve) [inline .dockerignore]: remove

File: .github/workflows/rust-cd.yml
Comment (zk-steve) [inline .github/workflows/rust-cd.yml:1]: review this file

File: .idea/.gitignore
Comment (zk-steve) [inline .idea/.gitignore:1]: don't commit this folder

File: Cargo.toml
Comment (zk-steve) [inline Cargo.toml:1]: use https://taplo.tamasfe.dev to sort the dependencies

File: assets/migrationJson/companies.json
Comment (zk-steve) [inline assets/migrationJson/companies.json:1]: empty file ?

File: assets/migrationJson/jobs.json
Comment (zk-steve) [inline assets/migrationJson/jobs.json:1]: empty file ?

File: assets/migrationJson/map_resume_job.json
Comment (zk-steve) [inline assets/migrationJson/map_resume_job.json:1]: empty file ?

File: assets/migrationJson/resumes.json
Comment (zk-steve) [inline assets/migrationJson/resumes.json:1]: empty file ?

File: config-default-docker.toml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+log_level="info"
+service_name="rust-api-service"
+#default value in project is postgres database
+#if database none or database: "in-memory" to set value is in-memory
```
Comment (zk-steve) [inline config-default-docker.toml]: grammar error

File: migrations/20240729112756_users_table.up.sql [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+values ('hr3@gmail.com', '$argon2i$v=19$m=4096,t=3,p=1$x+BlYXdrM9ndT+P42TYNt7UNOnn6xslz75HzSjY9+A8$iIXBtO+glnyIT27QjDilaoTNYTNsYq38M6gzK3XqpxE', 1, 3, false);
+
+INSERT INTO users (email, password, company_id, role_id, is_delete)
+values ('user1@gmail.com', '$argon2i$v=19$m=4096,t=3,p=1$x+BlYXdrM9ndT+P42TYNt7UNOnn6xslz75HzSjY9+A8$iIXBtO+glnyIT27QjDilaoTNYTNsYq38M6gzK3XqpxE', 0, 2, false);
+
+INSERT INTO users (email, password, company_id, role_id, is_delete)
+values ('user2@gmail.com', '$argon2i$v=19$m=4096,t=3,p=1$x+BlYXdrM9ndT+P42TYNt7UNOnn6xslz75HzSjY9+A8$iIXBtO+glnyIT27QjDilaoTNYTNsYq38M6gzK3XqpxE', 0, 2, false);
+
+INSERT INTO users (email, password, company_id, role_id, is_delete)
+values ('user3@gmail.com', '$argon2i$v=19$m=4096,t=3,p=1$x+BlYXdrM9ndT+P42TYNt7UNOnn6xslz75HzSjY9+A8$iIXBtO+glnyIT27QjDilaoTNYTNsYq38M6gzK3XqpxE', 0, 2, false);
```
Comment (zk-steve) [inline migrations/20240729112756_users_table.up.sql]: Don't inject dummy data into database (avoid production)

File: migrations/20240802113719_companies_table.up.sql [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    updated_on TIMESTAMP NOT NULL DEFAULT NOW()
+);
+
+INSERT INTO companies (name, email, address, description, is_delete)
+values ('123', '12next@gmail.com', 'So 4 Nguyen Trai', 'Web', false);
+
+INSERT INTO companies (name, email, address, description, is_delete)
+values ('1234', '1234next@gmail.com', 'So 4 Nguyen Trai', 'Web', false);
+
+INSERT INTO companies (name, email, address, description, is_delete)
```
Comment (zk-steve) [inline migrations/20240802113719_companies_table.up.sql]: .

File: migrations/20240802122842_resumes_table.up.sql [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+INSERT INTO resumes (user_id, email, url, is_delete)
+values (7 , 'user2@gmail.com', 'abcxyz', false);
+
+INSERT INTO resumes (user_id, email, url, is_delete)
+values (6, 'user1@gmail.com', 'abcxyz', false);
+
+INSERT INTO resumes (user_id, email, url, is_delete)
+values (6, 'user1@gmail.com', 'abcxyz', true);
+
```
Comment (zk-steve) [inline migrations/20240802122842_resumes_table.up.sql]: .

File: migrations/20240806031847_job_table.up.sql [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+values ('Intern Smart Contract', '1', 'So 2 Pham Van Bach', 4, 2000000, 'Intern' , 'Intern Smart Contract', false);
+
+INSERT INTO jobs (job_name, company_id, location, quantity, salary, job_level, description, is_delete)
+values ('Junior Smart Contract', '1', 'So 2 Pham Van Bach', 3, 8000000, 'Junior', 'Junior Smart Contract', false);
+
+INSERT INTO jobs (job_name, company_id, location, quantity, salary, job_level, description, is_delete)
+values ('BA Smart Contract', '2', 'So 4 Nguyen Trai', 2, 15000000, 'Junior', 'BA Smart Contract', false);
+
+INSERT INTO jobs (job_name, company_id, location, quantity, salary, job_level, description, is_delete)
+values ('Senior Smart Contract', '2', 'So 4 Nguyen Trai', 2, 25000000, 'Senior', 'Senior Smart Contract', false);
```
Comment (zk-steve) [inline migrations/20240806031847_job_table.up.sql]: .

File: migrations/20240808044729_map_resume_job_table.up.sql [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+INSERT INTO map_resume_job (resume_id, job_id)
+values (1, 2);
+
+INSERT INTO map_resume_job (resume_id, job_id)
+values (2, 2);
+
+INSERT INTO map_resume_job (resume_id, job_id)
+values (3, 2);
+
+INSERT INTO map_resume_job (resume_id, job_id)
```
Comment (zk-steve) [inline migrations/20240808044729_map_resume_job_table.up.sql]: ..

File: src/.DS_Store
Comment (zk-steve) [inline src/.DS_Store:1]: .

File: src/configs/config.rs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    pub host: String,
+    // Which PORT the server is listening to
+    pub port: u16,
+    // Which PORT the server jaeger collection is listening to
+    pub jaeger_port: u16,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct Config {
+    // Which errors we want to log (info, warn or error)
```
Comment (zk-steve) [inline src/configs/config.rs]: "Which errors we want to log (info, warn or error)" is not true for log_level

File: src/controllers/company.rs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+use crate::models::pagination::{Pagination, PaginationMethods};
+use crate::models::store_trait::StoreMethods;
+use crate::service::convert_to_json::{Data, PayloadNoData, PayloadWithData};
+use crate::service::jwt::Claims;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tracing::{event, instrument, Level};
+use warp::http::StatusCode;
+
+// Handle for create company
```
Comment (zk-steve) [inline src/controllers/company.rs]: check grammar

File: src/controllers/company.rs [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+//
+// This function adds a new company to the system. It takes company information to
+// be created and a reference to the StoreMethods trait object for company. It
```
Comment (zk-steve) [inline src/controllers/company.rs:14]: "It takes company information to be created and a reference to the StoreMethods trait object for company. " No helpful info

File: src/controllers/company.rs [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+// returns a success response with status 200 if company is created successfully
+#[instrument(level = "info", skip(store))]
+pub async fn create_company(
+    store: Arc<dyn StoreMethods + Send + Sync>,
+    claims: Claims,
+    new_company: NewCompany,
+) -> Result<impl warp::Reply, warp::Rejection> {
+    //Check valid company
+    let new_email = new_company.email.clone();
+    if let Ok(_res) = store.get_company_by_email(new_email).await {
```
Comment (zk-steve) [inline src/controllers/company.rs:24]: get_company_by_email should take &str instead of String

File: src/controllers/company.rs [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+        let payload = PayloadNoData {
+            message: "Email company already exists".to_string(),
+        };
+        return Ok(warp::reply::with_status(
+            warp::reply::json(&payload),
+            StatusCode::BAD_REQUEST,
+        ));
+    }
+
+    let company = NewCompany {
```
Comment (zk-steve) [inline src/controllers/company.rs]: why not use new_company directly?

File: src/controllers/company.rs [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+#[instrument(level = "info", skip(store))]
+pub async fn get_list_company(
+    store: Arc<dyn StoreMethods + Send + Sync>,
+    params: HashMap<String, String>,
+) -> Result<impl warp::Reply, warp::Rejection> {
+    // Get pagination from query parameters
+    let mut pagination = Pagination::default();
+
+    if !params.is_empty() {
+        event!(Level::INFO, pagination = true);
```
Comment (zk-steve) [inline src/controllers/company.rs]: why logging this?

File: src/controllers/company.rs [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+        pagination = <Pagination as PaginationMethods>::extract_pagination(params)?;
+    }
+    // Get list companies with pagination filters
+    match store
+        .get_list_company(pagination.limit, pagination.offset)
+        .await
+    {
+        Ok(res) => {
+            let payload = PayloadWithData {
+                message: "Get List Company Success".to_string(),
```
Comment (zk-steve) [inline src/controllers/company.rs]: message normally only returns "success"

File: src/controllers/job.rs [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+// returns a success response with status 200 if job is created successfully
+#[instrument(level = "info", skip(store))]
+pub async fn create_job(
+    store: Arc<dyn StoreMethods + Send + Sync>,
+    claims: Claims,
+    new_job: NewJob,
+) -> Result<impl warp::Reply, warp::Rejection> {
+    // Check authorization create job of the user
+    if let Ok(res) = store.get_user_by_id(claims.id).await {
+        if res.company_id != new_job.clone().company_id {
```
Comment (zk-steve) [inline src/controllers/job.rs]:
```suggestion
        if res.company_id != new_job.company_id.clone() {
```

Comment (zk-steve) [reply src/controllers/job.rs]: better compare &res.company_id

File: src/controllers/job.rs [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+            let payload = PayloadNoData {
+                message: "Un authorization create job".to_string(),
+            };
+            return Ok(warp::reply::with_status(
+                warp::reply::json(&payload),
+                StatusCode::BAD_REQUEST,
+            ));
+        }
+    }
+    let job = NewJob {
```
Comment (zk-steve) [inline src/controllers/job.rs]: .

File: src/controllers/job.rs [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+// be created and a reference to the StoreMethods trait object for job. It
+// returns a success response with status 200 if job is created successfully
+#[instrument(level = "info", skip(store))]
+pub async fn create_job(
+    store: Arc<dyn StoreMethods + Send + Sync>,
+    claims: Claims,
+    new_job: NewJob,
+) -> Result<impl warp::Reply, warp::Rejection> {
+    // Check authorization create job of the user
+    if let Ok(res) = store.get_user_by_id(claims.id).await {
```
Comment (zk-steve) [inline src/controllers/job.rs:24]: let Ok(res) = store.get_user_by_id(claims.id).await?

Comment (zk-steve) [inline src/controllers/job.rs]:
```suggestion
    let res = store.get_user_by_id(claims.id).await?;
```

File: src/controllers/user.rs [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+            warp::reply::json(&payload),
+            StatusCode::BAD_REQUEST,
+        ));
+    }
+    let hash_password = hash_password(new_user.password.as_bytes());
+    let user = AuthInfo {
+        email: new_email,
+        password: hash_password,
+    };
+    match store.create_user(user).await {
```
Comment (zk-steve) [inline src/controllers/user.rs:59]: never save password to db

File: src/service/convert_to_json.rs
Comment (zk-steve) [inline src/service/convert_to_json.rs:1]: name the folder to lib or something else. it's not `service`

File: src/routes/user.rs [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+        .or(set_hr_api)
+        .or(update_admin_api)
+        .or(update_admin_password_api)
+        .or(delete_admin_api)
+        .or(set_admin_api)
+}
+//test
+// #[cfg(test)]
+// #[path = "../tests/route_user.rs"]
+// mod route_user_tests;
```
Comment (zk-steve) [inline src/routes/user.rs]: remove

File: .gitignore (hunk 1 above)
Comment (andrew2003) [reply .gitignore]: Done

File: .dockerignore (hunk 2 above)
Comment (andrew2003) [reply .dockerignore]: Done

File: Cargo.toml
Comment (andrew2003) [reply Cargo.toml:1]: Done

File: assets/migrationJson/companies.json
Comment (andrew2003) [reply assets/migrationJson/companies.json:1]: Remove Done

File: config-default-docker.toml (hunk 3 above)
Comment (andrew2003) [reply config-default-docker.toml]: Repair Done

**Comment by zk-steve** [comment]
Fill this checklist

# Check List

- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] Check every test passed.
- [x] Did you split imports into std, third parties, and custom parts?
- [x] Format your code.
  - [x] cargo fmt
  - [x] cargo clippy
  - [x] `taplo fmt --config taplo/taplo.toml`
- [x] Check unwrap, expect().
- [x] Check clone().
- [x] Check your commit messages.
- [x] Have you added meaningful comments.

File: src/controllers/company.rs (hunk 13 above)
Comment (andrew2003) [reply src/controllers/company.rs]: Done

File: src/controllers/company.rs (hunk 15 above)
Comment (andrew2003) [reply src/controllers/company.rs]: Done

File: src/controllers/job.rs (hunk 16 above)
Comment (andrew2003) [reply src/controllers/job.rs]: Done

File: src/controllers/job.rs (hunk 17 above)
Comment (andrew2003) [reply src/controllers/job.rs]: Done

File: src/controllers/job.rs (hunk 18 above)
Comment (andrew2003) [reply src/controllers/job.rs]: Done

File: src/service/convert_to_json.rs
Comment (andrew2003) [reply src/service/convert_to_json.rs:1]: I move to utils

File: src/configs/config.rs (hunk 9 above)
Comment (andrew2003) [reply src/configs/config.rs]: Done

File: src/controllers/company.rs (hunk 10 above)
Comment (andrew2003) [reply src/controllers/company.rs]: Done

File: src/routes/user.rs (hunk 20 above)
Comment (andrew2003) [reply src/routes/user.rs]: Done

File: migrations/20240729112756_users_table.up.sql (hunk 4 above)
Comment (andrew2003) [reply migrations/20240729112756_users_table.up.sql]: Done

File: .github/workflows/rust-clippy.yml
Comment (zk-steve) [inline .github/workflows/rust-clippy.yml:1]: recheck

File: .github/workflows/rust-ci.yml
Comment (zk-steve) [inline .github/workflows/rust-ci.yml:1]: recheck

File: src/controllers/user.rs [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+            Ok(warp::reply::with_status(
+                warp::reply::json(&payload),
+                StatusCode::OK,
+            ))
+        }
+        Err(e) => Err(warp::reject::custom(e)),
+    }
+}
+
+// 2 function handler for updating role of user.
```
Comment (zk-steve) [inline src/controllers/user.rs:249]: 2 ?

File: .github/workflows/rust-cd.yml
Comment (andrew2003) [reply .github/workflows/rust-cd.yml:1]: Done

File: .github/workflows/rust-ci.yml
Comment (andrew2003) [reply .github/workflows/rust-ci.yml:1]: Done

File: .github/workflows/rust-cd.yml [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+      DOCKER_HUB_REPO: andrewtran2801/rust-api-service
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Login to Docker Hub
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      - name: Build and push docker image
```
Comment (zk-steve) [inline .github/workflows/rust-cd.yml:22]:
```suggestion
```

**Pull request closed without merging**